src/EventStore.Core/TransactionLog/Scavenging/Stages/Calculator.cs

```diff
@@ -0,0 +1,286 @@
+
+					if (newStatus == originalStreamData.Status &&
+						newDiscardPoint == originalStreamData.DiscardPoint &&
+						newMaybeDiscardPoint == originalStreamData.MaybeDiscardPoint) {
+
+						// nothing to update for this stream
+					} else {
+						state.SetOriginalStreamDiscardPoints(
+							streamHandle: originalStreamHandle,
+							status: newStatus,
+							discardPoint: newDiscardPoint,
+							maybeDiscardPoint: newMaybeDiscardPoint);
+					}
+
+					totalCounter++;
+
+					// Checkpoint occasionally
+					if (++checkpointCounter == _checkpointPeriod) {
+						checkpointCounter = 0;
+						weights.Flush();
+						transaction.Commit(new ScavengeCheckpoint.Calculating<TStreamId>(
+							scavengePoint,
+							originalStreamHandle));
+						transaction = state.BeginTransaction();
+
+						var elapsed = stopwatch.Elapsed;
+						LogRate("period", _checkpointPeriod, elapsed - periodStart);
+						LogRate("total", totalCounter, elapsed);
+
+						periodStart = stopwatch.Elapsed;
+					}
+				}
+
+				// if we processed some streams, the last one is in the calculator
+				// if we didn't process any streams, the calculator contains the default
+				// none handle, which is probably appropriate to commit in that case
+				weights.Flush();
+				transaction.Commit(new ScavengeCheckpoint.Calculating<TStreamId>(
+					scavengePoint,
+					streamCalc.OriginalStreamHandle));
+				LogRate("grand total (including rests)", totalCounter, stopwatch.Elapsed);
+			} catch {
+				// invariant: there is always an open transaction whenever an exception can be thrown
+				transaction.Rollback();
+				throw;
+			}
+			_throttle.Rest(cancellationToken);
+		}
+
+		private void LogRate(string name, int count, TimeSpan elapsed) {
+			var rate = count / elapsed.TotalSeconds;
+			Log.Trace(
+				"SCAVENGING: Calculated in " + name + ": {count:N0} streams in {elapsed}. {rate:N2} streams per second",
+				count, elapsed, rate);
+		}
+
+		// This does two things.
+		// 1. Calculates and returns the discard points for this stream
+		// 2. Adds weight to the affected chunks so that they get scavenged.
+		//
+		// The calculator determines that we can definitely discard everything up to the discardPoint
+		// and we may be able to discard things between the discardPoint and the maybeDiscardPoint.
+		//
+		// We want to calculate the discard points from scratch, without considering what values they
+		// came out as last time.
+		private void CalculateDiscardPointsForOriginalStream(
+			EventCalculator<TStreamId> eventCalc,
+			WeightAccumulator weights,
+			StreamHandle<TStreamId> originalStreamHandle,
+			ScavengePoint scavengePoint,
+			CancellationToken cancellationToken,
+			ref int cancellationCheckCounter,
+			out DiscardPoint discardPoint,
+			out DiscardPoint maybeDiscardPoint) {
+
+			var fromEventNumber = 0L;
+
+			discardPoint = DiscardPoint.KeepAll;
+			maybeDiscardPoint = DiscardPoint.KeepAll;
+
+			//qq tuning: what would be sensible? probably pretty large
```

Comment:
maybe 200 to read at least one full logical block from disk when reading index files to avoid reading the same page twice (4kb / 24 bytes per index entry ~= 171). shouldn't result in a drastic performance boost though

Reply:
i set it to 256